### ISSUE TITLE
Remat Policy `minimal` does not work for MoE models

### ISSUE THREAD

**Issue by kocchop** [AI-Hypercomputer/maxtext#2207]
### Bug report

The remat policy `minimal` which essentially uses `jax.checkpoint_policies.checkpoint_dots_with_no_batch_dims` does not work with the expert gemms in MoE models because the expert gemms have einsum notation that has the `batch_dims`. One possible workaround is to use the remat policy `custom` and then manually specifying the different checkpoint names such as `mlpwi_0=device mlpwi_1=device mlpwo=device`.

For dense models, the `minimal` policy works.

<img width="1140" height="219" alt="Image" src="https://github.com/user-attachments/assets/985576ea-6586-4fea-9cf4-9f4af8bf7e2c" />

As we can see, for most part the `custom` remat policy with explicit checkpoint names' memory allocation requirement matches with defined remat policies except for `minimal` for the `mixtral-8x7b` case.

**Comment by RissyRan** [comment]
Thanks for reporting the issue! Do you have repro commands to share? For instance, in the MoE, what's the values for `sparse_matmul`, `megablox`, `capacity_factor` etc?

**Comment by gobbleturk** [comment]
ya this makes sense, thanks for filing! I think we should either remove minimal as an option, or change it to actually do the minimal remat by making use of "custom" under the hood and setting everything to device

**Comment by kocchop** [comment]
> Thanks for reporting the issue! Do you have repro commands to share? For instance, in the MoE, what's the values for `sparse_matmul`, `megablox`, `capacity_factor` etc?

Hi @RissyRan I ran on 8xGPUs with FSDP=1 and EP=8 parallelism config and `sparse_matmul=false megablox=false capacity_factor=1.0`

For defined policy used `remat_policy=minimal_flash`
Equivalent with custom would be `remat_policy=custom query_proj=device key_proj=device value_proj=device context=device out_proj=device mlpwi_0=device mlpwi_1=device mlpwo=device`

**Comment by RissyRan** [comment]
> ya this makes sense, thanks for filing! I think we should either remove minimal as an option, or change it to actually do the minimal remat by making use of "custom" under the hood and setting everything to device

```
      if cfg.remat_policy == "minimal":
        policy = jax.checkpoint_policies.checkpoint_dots_with_no_batch_dims
```

I see it's still supported in JAX [gradient-checkpointing](https://docs.jax.dev/en/latest/gradient-checkpointing.html). @gobbleturk do you see the cause that why Dense model works but not MoE? I didn't recall anything special in MLP block for dense.

**Comment by gobbleturk** [comment]
ya the OP describes why minimal doesn't work here, its because there is a "batch" dimension for moe matuls, where "batch" here means appears in both inputs and output (e.g. left hand side (LHS) , ride hand side (RHS), and output). The "batch" dimension for moe is actually the expert dimension since it appears in all of LHS, RHS and output:

moe matmul:
'BXE,XEM -> BXM` Here X is a "batch" dimension since it appears in all three (lhs, rhs, and output)

non-moe matul:
'BE, EM -> BM` Here there are no "batch dimensions" since there is no dimension that appears in all 3

Unfortunately the word "batch" is overloaded - the definition we care about here is if there is a dimension that appears in all three  einsum tensors (lhs, rhs and output). For Moe Matmuls the expert dimension satisfies this- it is a batch dimension and thus won't be saved (although I don't exactly follow the code [here](https://github.com/jax-ml/jax/blob/f724944c4200dbab913540a73202edcb9be59b8c/jax/_src/ad_checkpoint.py#L77))

we should either remove minimal, or better change our implementation of minimal from [checkpoint_dots_with_no_batch_dims](https://github.com/AI-Hypercomputer/maxtext/blob/59f9e06e65406eefb7e07dda8c390a07137cefef/MaxText/layers/decoders.py#L267-L268) to use our "custom" under the hood with every tensor explicitly listed out as "device"

**Comment by RissyRan** [comment]
Cool! Makes sense.

```
# one example for capped strategy

        dispatch_eimsum = "BSM,BSEC -> EBCM" --> B dim
        mlp_up_einsum = "EBCM,EMH -> EBCH" --> E dim
        mlp_down_einsum = "EBCH,EHM -> EBCM" --> E dim
        output_einsum = "EBCM,BSEC -> BSM" --> B dim
```

We could just update the minimal here for discrepancy between dense and moe. Will draft a PR shortly.
